tests/test_smc.py

```diff
@@ -0,0 +1,197 @@
+
+    # random weights: ESS should be between 1 and n_particles
+    n_instances = 100
+    for _ in range(n_instances):
+        n_particles = np.random.randint(1, 100)
+        log_weights = generate_log_weights(n_particles)
+        ess = effective_sample_size(log_weights)
+
+        assert ess >= 1
+        assert ess <= n_particles
+
+    # constant weights: ESS should be n_particles
+    for _ in range(n_instances):
+        n_particles = np.random.randint(1, 100)
+        constant = np.random.randn()
+        log_weights = np.zeros(n_particles) + constant
+        ess = effective_sample_size(log_weights)
+
+        np.testing.assert_almost_equal(ess, n_particles)
+
+    # degenerate weights: ESS should be 1
+    for _ in range(n_instances):
+        n_particles = np.random.randint(1, 100)
+        constant = np.random.randn()
+        index = np.random.randint(n_particles)
+        log_weights = np.log(np.zeros(n_particles))
+        log_weights[index] = constant
+
+        ess = effective_sample_size(log_weights)
+
+        np.testing.assert_almost_equal(ess, 1)
+
+
+@pytest.mark.parametrize("resampling_fxn", [identity_resample, multinomial_resample, conditional_multinomial_resample])
+def test_sequential_monte_carlo(resampling_fxn: Resampler):
+    """Run SMC with the desired resampling_fxn on a Gaussian 1D test system, and assert that:
+    * running estimates of the free energy as a fxn of lambda match analytical free energies, and
+    * endstate samples have expected mean and stddev
+    """
+    np.random.seed(2022)
+
+    # construct test system
+    target_mean, target_log_sigma = 1, -2
+    params = np.array([target_mean, target_log_sigma])  # TODO: randomize / parameterize
+    n_particles = 10000
+    n_windows = 100
+    lambdas = np.linspace(0, 1, n_windows)
+    u_fxn, _, sample, reduced_free_energy = make_gaussian_testsystem()
+    ref_free_energies = np.array([reduced_free_energy(lam, params) for lam in lambdas])
+    ref_delta_fs = ref_free_energies - ref_free_energies[0]
+
+    # prepare inputs for SMC
+    samples = sample(0, params, n_particles).flatten()
+    assert samples.shape == (n_particles,)
+
+    vec_u_fxn = jit(vmap(u_fxn, in_axes=(0, None, None)))
+
+    def log_prob(xs, lam):
+        xs = np.array(xs)
+        return -vec_u_fxn(xs, lam, params)
+
+    def propagate(xs, lam):
+        """random-walk metropolis with N(0, proposal_scale(lam)) proposals"""
+        n = len(xs)
+        xs = np.array(xs)
+        assert xs.shape == (n,)
+
+        proposal_scale = np.exp(lam * target_log_sigma)  # to match annealed_gaussian_def  # TODO: extract
+        proposals = xs + proposal_scale * np.random.randn(n)
+        assert proposals.shape == xs.shape
+
+        # metropolis
+        log_probs_before = log_prob(xs, lam)
+        log_probs_after = log_prob(proposals, lam)
+        log_accept_probs = np.minimum(log_probs_after - log_probs_before, 0)
+        accept_probs = np.exp(log_accept_probs)
```

Comment:
nice! `exp(min(x,0)) = min(exp(x),exp(0)) = min(exp(x),1)` was used to seeing the right hand side equation in wiki and had to pause for a second